[3.x] Add Resource Prefix in Cluster Creation for IAM resources

himani2411 opened the pull request:
### Description of changes
Changing Cluster Stack Creation to add Path and Name Prefix when creating IAM Resources and Lambda Resources 
(Role, Policy and Instance Profile)
* Describe *what* you're changing and *why* you're doing these changes.


### Tests
* Unit Test cases to Check Slurm Policies, Instance Profiles,S3 Policies, Parallel Cluster Policies, Roles and Lambda Roles have Paths and Name prefix.
* Manual Tests to check unit test cases and revert to Original Role Names when No Resource Prefix is provided.



### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Add Length Validation for Resource Prefix

Add Test cases for Length Validation as Resource Prefix needs to be less than equal to 30 characters.

Commit:
Add method for separating path and name prefix from resource Prefix

Add method which separates and provides path and name prefix from the Resource prefix the user provides.
Add Unit Test cases to check the different format of Resource Prefix

Commit:
Add CAPABILITY_NAMED_IAM for Create and Update of Cluster

Add CAPABILITY_NAMED_IAM for create, update of the cluster.
Manually tested the change by creating and Updating a cluster.

Commit:
Add Resource Prefix in cluster IAM resource Creation

Add Path and Name prefix to all the IAM Resources and Lambda IAM Cleanup Resource( Role, Instance Profile, Policy).
Add Unit Test cases and YAML for addition of Resource Prefix.
Using Stack Name, Iam Resource Type, Region and Name to create a unique Hash.
Change add_lambda_cfn_role to give smaller Role Id if Name prefix is given.

Commit:
Check S3Access Bucket Policy has Name prefix

Add Test case to check Name Prefix has been appended to S3Access Policy when a S3 Bucket is provided in HeadNode and Queue IAM section of YAML.
Code linting change for IAM Validator to remote unnecessary condition
Code Linting Change in Hashing suggested use of sha256

Commit:
Code linting change for Cfn.py

Code liniting and auto formating Cfn.py

Commit:
Change Folder name for Test cases

Change the Folder name to reflect the test function name

Commit:
Change Test cases for Resource Prefix

Remove Code Duplication for Test cases of Cluster creation and Template Generation

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,59 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+    full_resource_path = None
+    full_resource_name = None
+    if config.iam and config.iam.resource_prefix:
+        iam_path, iam_name_prefix = get_path_n_name_prefix_from_iam_resource_prefix(config.iam.resource_prefix)
```
it might be easier to just call this function `split_resource_prefix`

himani2411 replied:
Sure

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,59 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+        iam_path, iam_name_prefix = get_path_n_name_prefix_from_iam_resource_prefix(config.iam.resource_prefix)
+        if iam_name_prefix:
+            if iam_type in ["AWS::IAM::Role", "AWS::IAM::InstanceProfile"]:
+                full_resource_name = iam_name_prefix + name + "-"
```
You should break this up to get all the individual pieces first and assign it once to `full_resource_name` e.g. 

```
part_a = x
part_b = y
part_c = z + a + f
full_name = part_a + part_b + part_c
```

I think this will be easier to digest

himani2411 replied:
Made it more readable friendly

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,59 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+                    .capitalize()
+                )
+
+            elif iam_type == "AWS::IAM::Policy":
```
Why are we doing this for Policies?  It seems odd not to specify a name for the policy.

himani2411 replied:
Made the naming convention for all the IAM resources similar, rather than differentiating using Type

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,59 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+    if policy_name_prefix:
+        policy_name = policy_name_prefix + "-" + policy_name
+    if role_name:
+        role_id = f"{function_id}Role"
```
instead of reassigning the `role_id` we should only assign a value once with an if/else statement

himani2411 replied:
Assigning the value just once

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,59 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+        scope.config.cluster_name, scope.config, None, iam_type="AWS::IAM::Policy"
+    )
+    if role_path:
+        lambda_path = role_path
```
instead of reassigning the `lambda_path` we should only assign a value once with an if/else statement

himani2411 replied:
Assigning the value just once

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,59 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+    if role_path:
+        lambda_path = role_path
+    if policy_name_prefix:
+        policy_name = policy_name_prefix + "-" + policy_name
```
instead of reassigning the `policy_name` we should only assign a value once with an if/else statement

dreambeyondorange replied:
Also, why are we assigning the full policy name here instead of in the function `add_cluster_iam_resource_prefix`?

himani2411 replied:
Assigning the value just once. And changed the implementation to have similar naming convention for all IAM resources

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,59 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+    role_id = f"{function_id}FunctionExecutionRole"
+    lambda_path = IAM_ROLE_PATH
+    policy_name = "LambdaPolicy"
+    # Need to reduce calls to this function
```
Remove the comment if you aren't planning to do something about it in the PR.  You could also consider breaking the function up into two, since you only need half of it for the policy name

himani2411 replied:
Removed the comment. However, using the same function with same naming convention for all IAM resource

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -324,25 +367,44 @@ def _add_role_and_policies(self, node: Union[HeadNode, BaseQueue], name: str):
+        _, policy_name_prefix = add_cluster_iam_resource_prefix(
+            self._config.cluster_name, self._config, name, iam_type="AWS::IAM::Policy"
+        )
+        if policy_name_prefix:
```
use if/else to avoid reassigning the variable

himani2411 replied:
avoided reassigning the variable

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -397,12 +459,19 @@ def _add_s3_access_policies_to_role(self, node: Union[HeadNode, BaseQueue], role
+        _, policy_name_prefix = add_cluster_iam_resource_prefix(
+            self._config.cluster_name, self._config, name, iam_type="AWS::IAM::Policy"
+        )
+        if policy_name_prefix:
```
use if/else

himani2411 replied:
Assigning the value just once

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -422,8 +491,10 @@ def _add_s3_access_policies_to_role(self, node: Union[HeadNode, BaseQueue], role
 
-    def _cluster_scoped_iam_path(self):
+    def _cluster_scoped_iam_path(self, iam_path=None):
         """Return a path to be associated IAM roles and instance profiles."""
```
can you fix this comment?  `associated with IAM`

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -422,8 +491,10 @@ def _add_s3_access_policies_to_role(self, node: Union[HeadNode, BaseQueue], role
-    def _cluster_scoped_iam_path(self):
+    def _cluster_scoped_iam_path(self, iam_path=None):
         """Return a path to be associated IAM roles and instance profiles."""
+        if iam_path:
```
It's not required, but I think using if/else here is clearer

himani2411 replied:
Change the function to have a consistent styling

dreambeyondorange commented on cli/src/pcluster/templates/slurm_builder.py:
```diff
@@ -131,6 +134,8 @@ def _add_policies_to_compute_node_role(self, node_name, role):
             },
         ]
         policy_name = "parallelcluster-slurm-compute"
+        if policy_name_prefix:
```
please review all instances where you assign a variable and then do an if check to re-assign it to make it an if/else

himani2411 replied:
Assigning the value just once

dreambeyondorange commented on cli/src/pcluster/utils.py:
```diff
@@ -293,6 +293,23 @@ def get_resource_name_from_resource_arn(resource_arn):
+            return resource_prefix, None
+        elif resource_prefix.startswith("/") and not resource_prefix.endswith("/"):
+            # Contains Both
+            if resource_prefix.rfind("/") != -1:
```
we should write this so that we don't need so many if statements.  rfind will return -1 if there is no slash right?  and resource_prefix[:split_index] will return an empty string.  If its empty you can return None.  So I think it should be possible to do an rfind first and return the split result (assigning None when the split returns an empty string)

himani2411 replied:
Done used the rfind rather than using startswith and endswith

dreambeyondorange commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -79,6 +79,12 @@ def _validate(self, resource_prefix: str):
                 f"Please refer to our official documentation for further details.",
                 FailureLevel.ERROR,
             )
+        if len(resource_prefix) > 31:
```
why a hardcoded 31?  We should use shared variables to define the limits so that other modules can access the same info.

himani2411 replied:
I have used a constant a Limit of 30

dreambeyondorange replied:
What I mean is, this should be a variable defined in common.py or something similar and referenced from there instead of a hardcoded number here

dreambeyondorange commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -153,10 +153,21 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+            FailureLevel.ERROR,
+        ),
+        (
+            "/longerPath/pathprefix/nameprefix",
```
Why is the path included in the limit for the name prefix?  This path won't be reflected in the IAM names, so should not result in an error

himani2411 replied:
Thank you for point thing this.
I have change the logic and the error message to reflect the Length Limit on Name prefix.

dreambeyondorange commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -153,10 +153,21 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
             FailureLevel.ERROR,
         ),
+        (
+            "/path_.,+@=-prefix/name_.,+@=-prefix",
```
The path should not affect the name limit

himani2411 replied:
Thank you for point thing this.
I have change the logic and the error message to reflect the Length Limit on Name prefix.

dreambeyondorange reviewed (changes requested):
Great work Himani!  I left a few comments.

Commit:
Fix ResourcePrefix according to PR

Change Function get_path_n_name_prefix_from_iam_resource_prefix name to split_resource_prefix
Change implementation of split_resource_prefix, IamResourcePrefixValidator and add_cluster_iam_resource_prefix
Add constant IAM_NAME_PREFIX_LENGTH_LIMIT as 30
Change Multiple variable and Comments
Creating Hash without the Prefix and separating full resource name creation.

PR Link:
https://github.com/aws/aws-parallelcluster/pull/4548

Commit:
Change Hash generation for IAM resource

Keeping the Hash as Upper case
Change _cluster_scoped_iam_path t have consistent styling.

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,54 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+        lambda_path = role_path
+    else:
+        lambda_path = IAM_ROLE_PATH
+    if policy_name_prefix:
```
similar ternary option here

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -250,17 +251,54 @@ def get_queue_security_groups_full(managed_compute_security_group: ec2.CfnSecuri
+        policy_name = policy_name_prefix
+    else:
+        policy_name = "LambdaPolicy"
+    if role_name:
```
ternary as well

Commit:
Fix default values assignment for IAM resources

Using Short circuiting for assignment of default value assignment.

Commit:
Fix default values assignment for IAM resources

Using Short circuiting for assignment of default value assignment.

Commit:
Add Constant for Path Length Limit
in Resource Prefix

Add Constant for PATH Length Limit
Add Validation for Path length and Test case
Add correct message for Resource Prefix Validation Length Check
Add correct variable name for policy name

dreambeyondorange commented on cli/src/pcluster/constants.py:
```diff
@@ -200,3 +200,6 @@
 LAMBDA_VPC_ACCESS_MANAGED_POLICY = "arn:${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole"
+
+IAM_NAME_PREFIX_LENGTH_LIMIT = 30
+IAM_PATH_PREFIX_LENGTH_LIMIT = 512
```
I would probably just call this the path rather than the path prefix.  It's the whole path right?

himani2411 replied:
Sure

dreambeyondorange commented on cli/src/pcluster/templates/cdk_builder_utils.py:
```diff
@@ -396,13 +442,16 @@ def _add_s3_access_policies_to_role(self, node: Union[HeadNode, BaseQueue], role
             policy_document=iam.PolicyDocument(statements=[]),
             roles=[role_ref],
-            policy_name="S3Access",
+            policy_name=policy_name_prefix or "S3Access",
```
I wonder if we should be appending the prefix to the defaults if they exist?  Seems useful to always have `S3Access` in the name for example

himani2411 replied:
i will change the Variable name as it is no longer a prefix but full policy name that I am returning due to implementation change for add_cluster_iam_resource_prefix()

himani2411 replied:
Or do you mean not keeping the prefix to policy name at all.

dreambeyondorange commented on cli/src/pcluster/utils.py:
```diff
@@ -293,6 +293,24 @@ def get_resource_name_from_resource_arn(resource_arn):
     return resource_arn.rsplit("/", 1)[-1] if resource_arn else ""
 
 
+def split_resource_prefix(resource_prefix):
```
nice!

himani2411 replied:
Thanks

dreambeyondorange commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -79,6 +80,19 @@ def _validate(self, resource_prefix: str):
+            )
+        if iam_path_prefix and (len(iam_path_prefix) > IAM_PATH_PREFIX_LENGTH_LIMIT):
+            self._add_failure(
+                f"Length of Path {iam_path_prefix} should be upto {IAM_PATH_PREFIX_LENGTH_LIMIT} characters. "
```
`must be less than` instead of `should be upto`

himani2411 replied:
Done

himani2411 committed:
Merge branch 'develop' into wip/iam_resource_prefix

himani2411 committed:
Merge branch 'develop' into wip/iam_resource_prefix
